### DISCUSSION TITLE
Using sessionmaker as contextmanger forces to expunge manually

### DISCUSSION THREAD

**Discussion by cartisan** [sqlalchemy/sqlalchemy#8016] [Usage Questions]
### Describe the bug

I'm using SQLAlchemy and trying to use the sessionmaker as a context manager for my transaction (following the [documentation][1]):
```python
    Session = sessionmaker(some_engine)
    with Session.begin() as session:
        query = session.query(SomeTable)
        an_entry = query.one()
        # session.expunge_all()  # Fixes the exception
    an_entry.attribute
```
Which raises an `sqlalchemy.orm.exc.DetachedInstanceError: Instance <AnEntry at 0x7f0a9d1c2560> is not bound to a Session` exception. This can be solved by manually expunging inside the contextmanager (see the line commented out above).

I'm surprised by this behavior, as the [SQLAlchemy documentation][2] and code indicate that closing a session should automatically expunges all ORM objects: In fact, a breakpoint in the `expunge_all` method of class `Session`  in `orm.session.py` is triggered when the contextmanager in the code above is exited. The same breakpoint is also triggered in the manually call to `expunge_all` inside the contextmanager. I'm very confused by this fact. Why does the first expunge caused by exiting the contextmanager not work?

Interestingly, the following (taken from the [docs][3], too) works _without_ me having to manually expunge:
```python
    Session = sessionmaker(some_engine)
    session = Session()
    with session.begin():
        query = session.query(SomeTable)
        an_entry = query.one()
    an_entry.attribute
```

The difference seems to be that `Session.begin() ` uses `session._maker_context_manager()` to create the context manager, which looks like this:
```python
    def _maker_context_manager(self):
        with self:
            with self.begin():
                yield self
```
And indeed, taking this over into my setup reintroduces the initial error:
```python
    Session = sessionmaker(some_engine)
    session = Session()
    with session:   # <-- this is the culprit that forces me to expunge manually
        with session.begin():
            query = session.query(SomeTable)
            an_entry = query.one()
    an_entry.attribute
```
I'm confused by two things:
1. Is `session._maker_context_manager()` implemented correctly, or is the line containing `with self:` there by error?
2. If the `with self:` contextmanager is actually needed, why is the documentation [linked above][3] not making use of it?

  [1]: https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker.begin
  [2]: https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.close
  [3]: https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#explicit-begin

### To Reproduce

```python
Session = sessionmaker(some_engine)
    with Session.begin() as session:
        query = session.query(SomeTable)
        an_entry = query.one()
        # session.expunge_all()  # Fixes the exception
    an_entry.attribute
```

### Error

```
Traceback (most recent call last):
  File "/home/leonid/code/prism_ada/prism_ada/session_demo.py", line 75, in <module>
    model_slug = forecast_model_schema.slug
  File "/home/leonid/.local/share/virtualenvs/prism_ada-nMYNssaU/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 481, in __get__
    return self.impl.get(state, dict_)
  File "/home/leonid/.local/share/virtualenvs/prism_ada-nMYNssaU/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 926, in get
    value = self._fire_loader_callables(state, key, passive)
  File "/home/leonid/.local/share/virtualenvs/prism_ada-nMYNssaU/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 957, in _fire_loader_callables
    return state._load_expired(state, passive)
  File "/home/leonid/.local/share/virtualenvs/prism_ada-nMYNssaU/lib/python3.10/site-packages/sqlalchemy/orm/state.py", line 710, in _load_expired
    self.manager.expired_attribute_loader(self, toload, passive)
  File "/home/leonid/.local/share/virtualenvs/prism_ada-nMYNssaU/lib/python3.10/site-packages/sqlalchemy/orm/loading.py", line 1369, in load_scalar_attributes
    raise orm_exc.DetachedInstanceError(
sqlalchemy.orm.exc.DetachedInstanceError: Instance <ForecastModelSchema at 0x7fadf1fb2590> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)
``` 

### Versions

- OS: Ubuntu 21.10
- Python: 3.10
- SQLAlchemy: 1.4.27
- Database: Postgres

**Comment by zzzeek** [comment]
hi -

an important concept is that the commit operation expires all state within the objects.  This is introduced at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#session-committing .  So the behavior illustrated is expected.

As I went through the docs for this, coming from the particular information you had up front, I can see it was not that immediate to see this, so I will add some more notes at https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html which is where we're responsible for ensuring top-level concepts are introduced as well as at some of your links.

**Reply by zzzeek** [reply]
oh actually it is here: https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html#closing-a-session   in the tutorial

**Reply by cartisan** [reply]
Dear zzzeek, thanks for coming back to this so swiftly! :)

I think I still struggle connecting the dots. Does what you say mean that
```python
Session = sessionmaker(some_engine)
with Session.begin() as session:
    query = session.query(SomeTable)
    an_entry = query.one()
    session.expunge_all()  # Fixes the exception
an_entry.attribute
```
works, because I expunge before the closing of the session by the contextmanager expires `an_entry`? While simply closing the session through contextmanager is not intended to expunge ORM objects but only expire them? 

What should be the preferred way of doing this, when one needs to access attributes (read only) on ORM objects like `an_entry` outside the scope of the contextmanager? Like above, including the `expunge`, or rather like below and without any `expunge`:
```python
Session = sessionmaker(some_engine)
session = Session()
with session.begin():
    query = session.query(SomeTable)
    an_entry = query.one()
an_entry.attribute
```

**Reply by zzzeek** [reply]
> Dear zzzeek, thanks for coming back to this so swiftly! :)
> 
> I think I still struggle connecting the dots. Does what you say mean that
> 
> ```python
> Session = sessionmaker(some_engine)
> with Session.begin() as session:
>     query = session.query(SomeTable)
>     an_entry = query.one()
>     session.expunge_all()  # Fixes the exception
> an_entry.attribute
> ```
> 
> works, because I expunge before the closing of the session by the contextmanager expires `an_entry`? 

the context manager you're using is "begin()", which means at the end, it will "commit()", before closing.   That's the part that does the expiration.  So yes, think of it as "an_entry" jumping from a sinking ship :) .

> While simply closing the session through contextmanager is not intended to expunge ORM objects but only expire them?

the context manager you have above will do both because it's closing also.  

> 
> What should be the preferred way of doing this, when one needs to access attributes (read only) on ORM objects like `an_entry` outside the scope of the contextmanager? Like above, including the `expunge`, or rather like below and without any `expunge`:
> 
> ```python
> Session = sessionmaker(some_engine)
> session = Session()
> with session.begin():
>     query = session.query(SomeTable)
>     an_entry = query.one()
> an_entry.attribute
> ```

use expire_on_commit=False

```
Session = sessionmaker(some_engine, expire_on_commit=False)
```

**Reply by cartisan** [reply]
Thanks for the clarification!
